File: csharp/behavioral/rare_visitor.cs
namespace Gof.ModernCSharp.Behavioral;

// Visitor puts operations in a separate object that each element type can accept.
// Modern C# note: consider pattern matching or polymorphic methods first. Visitor
// helps when new operations must be added without changing element classes.
public static class RareVisitor
{
    interface IShape { double Accept(IShapeVisitor visitor); }
    sealed record Circle(double Radius) : IShape
    {
        public double Accept(IShapeVisitor visitor) => visitor.Visit(this);
    }
    sealed record Rectangle(double Width, double Height) : IShape
    {
        public double Accept(IShapeVisitor visitor) => visitor.Visit(this);
    }
    interface IShapeVisitor
    {
        double Visit(Circle circle);
        double Visit(Rectangle rectangle);
    }
    sealed class AreaVisitor : IShapeVisitor
    {
        public double Visit(Circle circle) => Math.PI * circle.Radius * circle.Radius;
        public double Visit(Rectangle rectangle) => rectangle.Width * rectangle.Height;
    }

    public static string Run()
    {
        IShape[] shapes = [new Circle(2), new Rectangle(3, 4)];
        return string.Join(", ", shapes.Select(shape => shape.Accept(new AreaVisitor()).ToString("0.##")));
    }
}

File: py/behavioral/rare_visitor.py
# Visitor puts operations in a separate object that each element type can accept.
# Modern Python note: consider pattern matching, singledispatch, or polymorphic
# methods first. Visitor helps when new operations must be added without changing
# the element classes.
class Circle:
    def __init__(self, radius):
        self.radius = radius

    def accept(self, visitor):
        return visitor.visit_circle(self)


class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def accept(self, visitor):
        return visitor.visit_rectangle(self)


class AreaVisitor:
    def visit_circle(self, circle):
        return 3.14 * circle.radius * circle.radius

    def visit_rectangle(self, rectangle):
        return rectangle.width * rectangle.height


if __name__ == "__main__":
    shapes = [Circle(2), Rectangle(3, 4)]
    print([shape.accept(AreaVisitor()) for shape in shapes])
